### ISSUE TITLE
Plan for migrating Kyma components to new modular arch 

### ISSUE THREAD

**Issue by arturskorupa** [kyma-project/kyma#15452]
## Description

Define principal rules for defining interfaces between modules are defined based on all potential use cases of current Kyma components.  

Kyma component teams has already everything to start creating their plans to migrate to the new modular architecture: they finished POC about transition to the new architecture to proof feasibility of the concept (also areas to be addressed by further development has been identified - list is kept [here](https://github.com/kyma-project/kyma/issues/15452#issuecomment-1254636824)).

**The goal of this task is to define a plan remaining work for each component.** (that will be linked in the relevant acceptance criteria item).
The plan should provide answers on the following questions:
- Is the considered module needed in Kyma
- Are there any blockers the component team foresee based on the current known deliverables from the central reconciler team.  What is the remedy (issue for the module team or central reconciler team is a must) 
- What date we can expect the module to become GA
- Are there any special requirements for the module (in terms of release management, governance etc.)

## Acceptance Criteria

- [x] Principal rules for defining interfaces between modules are defined based on all potential use cases of current Kyma components ([dependencies between components](https://github.com/kyma-project/community/tree/main/concepts/modularization#dependencies-between-components))
- [ ] Review overrides provided to `Kyma Reconciler` by KEB and decision is made how new modules are going to consume them (or not?).  (MOVED TO another issue #16246)
- [x] It is defined which of Kyma modules is available for internal and external usage 
- [x] Migration plan for Service mesh
- [x] Migration plan for API Gateway
- [x] Migration plan for Observability/Telemetry (https://github.tools.sap/huskies/knowledge-hub/blob/master/modularization/overview.md, https://github.tools.sap/huskies/knowledge-hub/tree/master/telemetry-module)
- [x] Migration plan for Eventing (https://github.tools.sap/skydivingtunas/proposals/blob/eventing-modularization/eventing-modularization/README.md)
- [x] Migration plan for Application Connector
- [x] Migration plan for Connectivity Proxy
- [x] Migration plan for Compass Runtime Agent
- [x] Migration plan for BTP Operator
- [x] Migration plan for SRE modules
- [x] No other components are left (like certificates, cluster-essentials etc.)

**Links**
- [Concept document](https://github.com/kyma-project/community/tree/main/concepts/modularization)
- [Transition plan ](https://github.com/kyma-project/community/blob/main/concepts/modularization/transition.md)
- [Lifecycle manager](https://github.com/kyma-project/lifecycle-manager)
- [Module manager](https://github.com/kyma-project/module-manager)
- [Template operator](https://github.com/kyma-project/template-operator)

**Comment by arturskorupa** [comment]
Tasks to perform to move component to modular architecture (not exhaustive list):
- create operator for module
- prepare UI
- refactor tests (could happen there's no common integration tests suite any more - #15442)
- detach from old reconciler and attach to the Lifecycle Manager
- documentation changes
- establish all needed build and governance pipelines

**Comment by arturskorupa** [comment]
> Review overrides provided to `Kyma Reconciler` by KEB and decision is made how new modules are going to consume them (or not?)
consider a new setting on a module level (provided when the module is getting enabled) e.g. evaluation / production profile

**Comment by arturskorupa** [comment]
> Principal rules for defining interfaces between modules are defined based on all potential use cases of current Kyma components

Missing dependencies has to be handled gracefully by the module-operator. There aren't any common patterns of interaction between components (in current Kyma version) that can be extracted as an interface or _principal rule_

**Comment by pbochynski** [comment]
Dependencies should be avoided and minimized. General guideline I am going to add to the modularization concept:
- module operator should know its dependencies
- keep dependencies minimal and try to depend on API only (check if API exists in the required version)
- report missing dependencies in the status of the module operator resource
- remember that dependency can be installed in parallel to your module - give it some time and report error after reasonable timeout
- error or success is not a final state (your dependencies can come and go) - reconcile, update status, bring it to the desired state eventually

**Comment by strekm** [comment]
as for @kyma-project/goat we started working on istio manager and that will be our focus at first. Istio manager got no dependencies, it will expose Istio CRD as API, proposal can be found [here](https://github.com/kyma-project/istio/blob/main/docs/xff-proposal.md). Istio manager will consist of component managing istio itself, component managing sidecars, component managing other resources (istio-resources, certificates). this will be described in technical design (TBD).

api-gateway module will be worked on later on since it simpler component. It has dependency to Istio CRDs. the will consist of api-gateway controller, ORY stack (as of now). API is not designed yet nor any POCs were done so far

**Comment by PK85** [comment]
BTP-Operator -> BTP-Manager

- Is the considered module needed in Kyma? _Answer: Yes, it's default module installed always when new SKR is created. We have a way to make it optional._
- Are there any blockers the component team foresee based on the current known deliverables from the central reconciler team. What is the remedy? _Answer: I found that problem and asked LM team to solve that: https://github.com/kyma-project/module-manager/issues/164  (this affects all modules)_
- What date we can expect the module to become GA? _Answer: Dec/January_
- Are there any special requirements for the module (in terms of release management, governance etc.)? ### Answer: This module requires to receive secret from central KCP cluster/KEB. Mitigation: KEB provides secret into the Kyma cluster.

**Comment by akgalwas** [comment]
**Migration plan for Application Connector and Compass Runtime Agent**
Application Connector components are tightly coupled with Compass Runtime Agent. After discussions it was decided that Application Connector module will contain the following components: Compass Runtime Agent, Connectivity Validator, and Application Gateway.
 

- Is the considered module needed in Kyma. A: Application Connectivity is a crucial element of Kyma cluster. It allows user workloads to invoke secured external APIs, and receive events. It can be used along with Eventing and Serverless modules. 
- Are there any blockers the component team foresee based on the current known deliverables from the central reconciler team. What is the remedy (issue for the module team or central reconciler team is a must). **A:** At this point there are no known blockers. 
- What date we can expect the module to become GA. **A:** end of Q1 2023 for the Application Connector module itself. As it comes to Compass Manager the expected date is Q3 2023. 
- Are there any special requirements for the module (in terms of release management, governance etc.). **A:** the module contains Compass Runtime Agent that needs configuration secret to be able to sync-up with Compass. In the current solution Provisioner component is responsible for providing the secret. There will be a new central component called Compass Manager introduced. It will be responsible for:
  -  watching Runtimes, and detecting Application Connector module installation
  - registering Runtimes in Compass
  - providing configuration secret for Compass Runtime Agent  

The module can be installed even if the secret is not delivered yet by Compass Manager.

**Comment by akgalwas** [comment]
**Migration plan for Connectivity Proxy**
Q&A:
- Is the considered module needed in Kyma. **A:** Connectivity Proxy has a big value for users.  
- Are there any blockers the component team foresee based on the current known deliverables from the central reconciler team. What is the remedy (issue for the module team or central reconciler team is a must). **A:** At this point there are no known blockers. 
- Are there any special requirements for the module (in terms of release management, governance etc.). **A:** Connectivity Proxy is a proprietary component and as such needs special treatment(e.g.storing in some internal repository).
- What date we can expect the module to become GA.

**Comment by arturskorupa** [comment]
> Review overrides provided to Kyma Reconciler by KEB and decision is made how new modules are going to consume them (or not?).

That topic is being discussed currently, still need some time to conclude.

**Comment by ebensom** [comment]
**Migration Plan for Runtime Monitoring Agent (RMA) and Runtime Monitoring Integration (RMI)**
 
- Is the considered module needed in Kyma: It is a module outside of Kyma, however it is a **mandatory** module on all SKRs (Kyma runtimes), without the possibility of the cluster users to disable the module.
- Are there any blockers the component team foresee based on the current known deliverables from the central reconciler team: No blockers at this point.
- What date we can expect the module to become GA: 2023.06.01.
- Are there any special requirements for the module (in terms of release management, governance etc.): Availability service level indicator (SLI) depends on Istio sidecar injection, and hence Istio module for these modules:
    - eventing (epp, )
    - application connector (connectivity validator, application gateway)
    - ory (hydra token API)
    - serverless (registry, webhook)


Migration efforts are tracked under 

- TBA

**Comment by arturskorupa** [comment]
> It is defined which of Kyma modules is available for internal and external usage
Modules used by customers (presented to them, with or without option to disable):
- Service mesh, 
- API Gateway
- Observability
- Telemtry
 -Eventing 
- Application Connector
- Connectivity Proxy


Modules for internal use (not presented to customers):
- Compass Manager
- BTP Operator
- SRE modules (RMI / RMA)

**Comment by arturskorupa** [comment]
Last open AC (Review overrides provided to Kyma Reconciler by KEB and decision is made how new modules are going to consume them (or not?). ) moved to new issue (https://github.com/kyma-project/kyma/issues/16246)
